### PR TITLE
[BEAM-455] Adds a test harnesses and utilities framework for sources.

### PR DESCRIPTION
**Pull request by chamikaramj**
Helper functions and test harnesses for checking correctness of source (``iobase.BoundedSource``) and range tracker (``iobase.RangeTracker``) implementations.

Contains a few lightweight utilities (e.g. reading items from a source such as ``readFromSource()``, as well as heavyweight property testing and stress testing harnesses that help getting a large amount of test
coverage with few code.

Most notable ones are:
* ``assertSourcesEqualReferenceSource()`` helps testing that the data read by the union of sources   produced by ``BoundedSource.split()`` is the same as data read by the original source.
* If your source implements dynamic work rebalancing, use the ``assertSplitAtFraction()`` family of functions - they test behavior of ``RangeTracker.try_split()``, in particular, that various consistency properties are respected and the total set of data read by the source is preserved when splits happen. 
Use ``assertSplitAtFractionBehavior()`` to test individual cases of dynamic work rebalancing and use ``assertSplitAtFractionExhaustive()`` as a heavy-weight stress test including concurrency.

Updates 'avroio_test.py' to use the test framework.

### CODE REVIEW

**Comment by chamikaramj** [comment]
R: @silviulica

File: sdks/python/apache_beam/io/source_test_utils.py [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+
+SplitFractionStatistics = namedtuple(
+    'SplitFractionStatistics',
+    'successful_fractions non_trivial_fractions')
+
+
+class SourceTestUtils(object):
+  """Helper functions and test harnesses for source implementations.
+
+  This module contains helper functions and test harnesses for checking
```
Comment (silviulica) [inline sdks/python/apache_beam/io/source_test_utils.py]: I notice this class has a large amount of static_methods. Why not have them as module level functions. should be the same. You actually seem to have this intention by the comment for the class saying "This module ...".

File: sdks/python/apache_beam/io/source_test_utils.py [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+  correctness of source (``iobase.BoundedSource``) and range tracker
```
Comment (silviulica) [inline sdks/python/apache_beam/io/source_test_utils.py]: say iobase.BoundedSource-derived?

Comment (chamikaramj) [inline sdks/python/apache_beam/io/source_test_utils.py]: Done.

File: sdks/python/apache_beam/io/source_test_utils.py (hunk 1 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/source_test_utils.py]: Yeah, this sounds better.

Moved class methods to be top level functions and moved the doc comment to the module level.

**Comment by chamikaramj** [comment]
Thanks Silviu. PTAL.

**Comment by silviulica** [comment]
LGTM - great work Cham!

**Comment by chamikaramj** [comment]
Thanks Silviu.

R: @dhalperi

**Comment by dhalperi** [comment]
Merged; please manually close PR

**Comment by chamikaramj** [comment]
Thanks Dan.

**Pull request closed without merging**